aiomisc/service/process.py

```diff
@@ -80,4 +92,34 @@ async def stop(self, exception: Exception = None) -> Any:
         await self.loop.run_in_executor(None, self.process_stop_event.wait)
 
 
-__all__ = ("ProcessService",)
+class RespawningProcessService(ProcessService, ABC):
+    process_poll_timeout: int = 5
+
+    _supervisor: PeriodicCallback
+
+    async def __supervise(self) -> None:
+        if not hasattr(self, "process"):
+            return
+
+        if await self.loop.run_in_executor(None, self.process.is_alive):
+            return
+
+        log.info(
+            "Process in service %r exited with code %r, respawning.",
+            self, self.process.exitcode,
+        )
+        await super().start()
+
+    async def start(self) -> None:
+        await super().start()
+        self._supervisor = PeriodicCallback(self.__supervise)
+        self._supervisor.start(
+            self.process_poll_timeout,
+        )
```

Comment:
You don't have to poll child processes. Just use `wait` system call.

Reply:
os.wait returns the first finished subprocess, but is not possible to detect what instance of WorkerPool is the owner of it.